core/imds/src/main/java/software/amazon/awssdk/imds/internal/DefaultEc2Metadata.java

```diff
@@ -152,41 +154,57 @@ public String toString() {
-            Optional<AbortableInputStream> responseBody = response.responseBody();
 
-            if (statusCode == HttpURLConnection.HTTP_OK && responseBody.isPresent()) {
-                abortableInputStream = responseBody.get();
-                data = IoUtils.toUtf8String(abortableInputStream);
-                metadataResponse = new MetadataResponse(data);
-            } else if (statusCode == HttpURLConnection.HTTP_NOT_FOUND) {
-                throw SdkServiceException.builder()
-                                         .message("The requested metadata at path ( " + path + " ) is not found ").build();
-            } else if (statusCode == HttpURLConnection.HTTP_OK) {
-                throw SdkClientException.builder()
-                                         .message("Response body empty with Status Code " + statusCode).build();
-            } else {
-                throw SdkClientException.builder()
-                                        .message("Instance metadata service returned unexpected status code " + statusCode)
-                                        .build();
+        for (int tries = 1 ; tries <= retryPolicy.numRetries() ; tries ++) {
+
+            try {
+                String token = getToken();
+                URI uri = URI.create(endpoint + path);
+                HttpExecuteRequest httpExecuteRequest = REQUEST_MARSHALLER.createDataRequest(uri, SdkHttpMethod.GET, token,
+                                                                                             tokenTtl);
+                HttpExecuteResponse response = httpClient.prepareRequest(httpExecuteRequest).call();
+                int statusCode = response.httpResponse().statusCode();
+                Optional<AbortableInputStream> responseBody = response.responseBody();
+
+                if (statusCode == HttpURLConnection.HTTP_OK && responseBody.isPresent()) {
+                    abortableInputStream = responseBody.get();
+                    String data = IoUtils.toUtf8String(abortableInputStream);
+                    metadataResponse = new MetadataResponse(data);
+                } else if (statusCode == HttpURLConnection.HTTP_NOT_FOUND) {
+                    throw SdkServiceException.builder()
+                                             .message("The requested metadata at path ( " + path + " ) is not found ").build();
+                } else if (statusCode == HttpURLConnection.HTTP_OK) {
+                    throw SdkClientException.builder()
+                                            .message("Response body empty with Status Code " + statusCode).build();
+                } else {
+                    throw SdkClientException.builder()
+                                            .message("Instance metadata service returned unexpected status code " + statusCode)
+                                            .build();
+                }
+            } catch (SdkServiceException sd) {
```

Comment:
Reiterating on the offline discussion, I have used SdkServiceException for status codes for which retries should not happen and SdkClientException for which retries are possible. In case of SdkServiceException , an error message is directly thrown , but for any other kind of exception it is caught and sent in the retry block.